api/v1alpha1/externaldns_types.go

```diff
@@ -18,41 +18,400 @@ package v1alpha1
+	//  * Infoblox
+	//
+	// +kubebuilder:validation:Required
+	// +unionDiscriminator
+	// +required
+	Type ExternalDNSProviderType `json:"type"`
+}
+
+// +kubebuilder:validation:Enum=AWS;GCP;Azure;BlueCat;Infoblox
+type ExternalDNSProviderType string
 
-	Spec   ExternalDNSSpec   `json:"spec,omitempty"`
-	Status ExternalDNSStatus `json:"status,omitempty"`
+const (
+	ProviderTypeAWS      ExternalDNSProviderType = "AWS"
+	ProviderTypeGCP      ExternalDNSProviderType = "GCP"
+	ProviderTypeAzure    ExternalDNSProviderType = "Azure"
+	ProviderTypeBlueCat  ExternalDNSProviderType = "BlueCat"
+	ProviderTypeInfoblox ExternalDNSProviderType = "Infoblox"
+	// More providers will ultimately be added in the future.
+)
+
+// ExternalDNSSource describes which Source resource
+// the ExternalDNS should create DNS records for.
+type ExternalDNSSource struct {
+	ExternalDNSSourceUnion `json:",inline"`
+
+	// HostnameAnnotationPolicy specifies whether or not ExternalDNS
+	// should ignore the "external-dns.alpha.kubernetes.io/hostname"
+	// annotation, which overrides DNS hostnames on a given source resource.
+	//
+	// The following values are accepted:
+	//
+	//  "Ignore": Ignore any hostname annotation overrides.
+	//  "Allow": Allow all hostname annotation overrides.
+	//
+	// The default behavior of the ExternalDNS is "Ignore".
+	//
+	// Note that by setting a HostnameAnnotationPolicy of "Allow",
+	// may grant privileged DNS permissions to under-privileged cluster
+	// users.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default:=Ignore
+	// +required
+	HostnameAnnotationPolicy *HostnameAnnotationPolicy `json:"hostnameAnnotation"`
+}
+
+// ExternalDNSSourceUnion describes optional fields for an ExternalDNS source that should
+// be captured.
+// +union
+type ExternalDNSSourceUnion struct {
+	// Type specifies an ExternalDNS source resource
+	// to create DNS records for.
+	//
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	// +required
+	Type ExternalDNSSourceType `json:"type"`
+
+	// AnnotationFilter describes an annotation filter
+	// used to filter which source instance resources
+	// ExternalDNS publishes records for.
+	// The annotation filter uses label selector semantics
+	// against source resource annotations.
+	//
+	// +optional
+	AnnotationFilter map[string]string `json:"annotationFilter,omitempty"`
+
+	// Namespace instructs ExternalDNS to only acknowledge
+	// source resource instances in a specific namespace.
+	//
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+
+	// Service describes source configuration options specific
+	// to the service source resource.
+	//
+	// +optional
+	Service *ExternalDNSServiceSourceOptions `json:"service,omitempty"`
+
+	// CRD describes source configuration options specific
+	// to the CRD source resource.
+	//
+	// +optional
+	CRD *ExternalDNSCRDSourceOptions `json:"crd,omitempty"`
+
+	// OpenShiftRoute source currently has no unique configuration options.
```

Comment:
@deads2k in your opinion, should we stub this type here, or leave it out like it is right now.

Reply:
> @deads2k in your opinion, should we stub this type here, or leave it out like it is right now.

I'm ok leaving it out for now.

Reply:
Thanks!